contracts/EpochTokenLocker.sol

```diff
@@ -0,0 +1,146 @@
+pragma solidity ^0.5.0;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/Math.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+// EpochTokenLocker saveguards tokens for applications with constant-balances during discrete epochs
+// It allows to deposit token which become credited in the next epoch and allows to request a token-withdraw
+// which becomes claimable after the current epoch expired.
+
+contract EpochTokenLocker {
+    using SafeMath for uint;
+
+    event Deposit(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event withdrawRequest(
+        address user,
+        address token,
+        uint amount,
+        uint stateIndex
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
```

Comment:
At some point we will need a way to iterate over all users in the system.

Reply:
In this implementation it would be only by using the events.
Is not enough? Or u mean it should have something to make it easier to iterate just with the smart contract:

```
mapping(address => uint) private userIds;
mapping(address => uint) private tokenIds;
address[] users;
address[] tokens;
```

Reply:
yes that exactly. I'd like to avoid having to maintain a second layer infrastructure.

Reply:
This will be implemented later in a separate PR. We want to use a set library, which is not yet implemented.